unix, windows: add uv_fs_event_start/stop functions

saghul opened the pull request:
Make uv_fs_event behave like other handles, that is, it's inactive after
init, and it's active between start and stop.

That.

I didn't add any new test (I plan to), but I'd like to get some feedback to see if I didn't fuck up anything too badly to begin with. Tests pass (I tested OSX, Linux and Windows).

The basic idea was to make init pretty dummy, put the meat of init now in start and what close did, do it in stop. But well, the devil is in the details :-)

Please review:

@bnoordhuis everything in general and nothing in particular :-)

@indutny please overview if the logic looks sound on OSX, where the fsevent thread is used

@piscisaureus my Windows-fu is not very strong, I'm trying! Please check if the stop process is done correctly, also the close. Basically I added some extra check not to fire any callback if there is an in-flight request when we called stop, in close use an endgame as usual. Also, I'm reusing handle->buffer until you close the handle, I hope that's ok.

saghul committed:
unix, windows: add uv_fs_event_start/stop functions

Make uv_fs_event behave like other handles, that is, it's inactive after
init, and it's active between start and stop.

Nodejs-Jenkins commented:
Thank you for contributing this pull request! Here are a few pointers to make sure your submission will be considered for inclusion.

Commit saghul/libuv@2eed05f26ca1b08c39573721447a9e895b558dd0 has the following error(s):
- First line of commit message must be no longer than 50 characters

You can fix all these things _without_ opening another issue.

Please see [CONTRIBUTING.md](https://github.com/joyent/node/blob/master/CONTRIBUTING.md) for more information

bnoordhuis commented on include/uv.h:
```diff
@@ -1880,8 +1877,15 @@ enum uv_fs_event_flags {
+UV_EXTERN int uv_fs_event_start(uv_fs_event_t* handle,
+                                uv_fs_event_cb cb,
+                                const char* filename,
+                                int flags);
```
s/int/unsigned int/ while you're here.

bnoordhuis commented on src/unix/kqueue.c:
```diff
@@ -296,23 +296,30 @@ static void uv__fs_event(uv_loop_t* loop, uv__io_t* w, unsigned int fflags) {
   int fd;
 
+  if (uv__is_active(handle))
+    return 0;
```
Maybe -EBUSY or -EINVAL?  I could live with turning this into an assert().

saghul replied:
I borrowed inspiration from fs_poll. I was actually thinking EALREADY for start and EINVAL for stop. This probably would need to be applied to other handles, I can do that afterwards, if that's ok.

bnoordhuis replied:
EBUSY or EINVAL is more POSIX-y.  I guess EINVAL for both start and stop is the least amount of mental overhead.

saghul replied:
This expands to all handles, so I'd leave fs_events like fs_poll is and then make a change adjusting the behavior globally, whatever we decide is best :-)

bnoordhuis commented on src/unix/kqueue.c:
```diff
@@ -335,13 +342,16 @@ int uv_fs_event_init(uv_loop_t* loop,
-void uv__fs_event_close(uv_fs_event_t* handle) {
+int uv_fs_event_stop(uv_fs_event_t* handle) {
+  if (!uv__is_active(handle))
+    return 0;
```
Ditto.

bnoordhuis commented on src/win/fs-event.c:
```diff
@@ -131,14 +107,42 @@ static int uv_split_path(const WCHAR* filename, WCHAR** dir,
+
+  uv_req_init(loop, (uv_req_t*)&handle->req);
+  handle->req.type = UV_FS_EVENT_REQ;
+  handle->req.data = (void*)handle;
```
Superfluous cast.

bnoordhuis commented on src/win/fs-event.c:
```diff
@@ -213,15 +217,17 @@ int uv_fs_event_init(uv_loop_t* loop, uv_fs_event_t* handle,
-    sizeof(DWORD));
+  if (!handle->buffer) {
+    handle->buffer = (char*)_aligned_malloc(uv_directory_watcher_buffer_size,
+                                            sizeof(DWORD));
```
Just curious, is the call to `_aligned_malloc()` necessary?  Put another way, can the normal malloc() return a pointer that's not at least sizeof(DWORD) aligned?

saghul replied:
@piscisaureus can tell you better :-)

bnoordhuis commented on src/win/fs-event.c:
```diff
@@ -456,22 +501,19 @@ void uv_process_fs_event_req(uv_loop_t* loop, uv_req_t* req,
 void uv_fs_event_endgame(uv_loop_t* loop, uv_fs_event_t* handle) {
-  if (handle->flags & UV__HANDLE_CLOSING &&
-      !handle->req_pending) {
+  if (handle->flags & UV__HANDLE_CLOSING && !handle->req_pending) {
```
Parentheses around first expression.

saghul commented:
Force pushed addressing Ben's review. I made start and stop return UV_EINVAL if called when the handle was already started / stopped respectively, I'll also add a test for that.

bnoordhuis commented:
LGTM from a code POV but it seems to break fs_event_watch_dir on OS X.  @piscisaureus should sign off on the Windows part, it probably wouldn't hurt for @indutny to take a peek at the FSEvents changes.

saghul commented:
Thanks for the review! The OSX test failed for me almost every time (on 10.6), FWIW. I'll run the tests (before and after) on 10.8 to see if there is any difference.

PS: What's the OSX version of the build bot?

tjfontaine commented:
It's on 10.8 atm

saghul commented:
Thanks @tjfontaine.

saghul commented:
Ok, forget what I said, looks like I was remembering wrong. I could indeed reproduce the OSX issue reliably on my 10.8 machine. I'll look into it.

saghul commented:
Ok, rebased, updated and fixed the OSX test failure, should be good to go now :-) @indutny can you please take a peak uv_fs_event_stop? I modified how the fsevents thread is stopped since now the handle is not closed, so I changed it to check if the handle is not active.

indutny commented:
@saghul OSX changes seems to be very minor and are generally looking good to me. LGTM

bnoordhuis commented:
Another LGTM.

saghul commented:
Thanks guys, landed on https://github.com/joyent/libuv/commit/9d44d786ada6cf94e1bdcee7d777c790ca712a78 :-)
